wc-api: Add user meta for UI preferences

coderkevin opened the pull request:
Fixes #892 

This adds wc-api code to store UI preferences and provides one example
of updating the revenue table to save column preferences.

### Detailed test instructions:

1. Go to Analytics -> Revenue.
2. Toggle some columns off.
3. Close browser tab.
4. Open another browser tab and go to Analytics -> Revenue (or use another browser, etc)
5. Observe that the same columns are toggled off.

For more detailed testing, open dev console network tab and look for "me?context=edit" requests and "me" POST requests.

coderkevin committed:
wc-api: Add user meta for UI preferences

This adds wc-api code to store UI preferences and provides one example
of updating the revenue table to save column preferences.

coderkevin committed:
wc-api user: fix returns of operations to arrays

Before, a function was returning undefined when the resourceNames didn't
include what we wanted for the user operations. This was causing some
console errors.

coderkevin committed:
ReportTable: use columnPrefsKey

This uses `columnPrefsKey` for the user preferences key instead of some
erroneously hard-coded values.

joshuatf commented on lib/admin.php:
```diff
@@ -338,5 +338,73 @@ function woocommerce_embed_page_header() {
+function wc_admin_get_user_data_values( $user ) {
+	$values = array();
+	foreach ( wc_admin_get_user_data_fields() as $field ) {
+		$values[ $field ] = get_user_meta( $user['id'], 'woocommerce_meta_' . $field, true );
```
Could you see anyone using this to get user meta not prefixed with `woocommerce_meta`?

Was thinking of using this as a 3rd party developer and it might be nice if we could add in other user meta fields:

```php
	$user_data_fields = array(
		'revenue_report_columns' => 'woocommerce_meta_revenue_report_columns',
         'my_custom_user_field' => 'other_plugin_prefix_meta_field',
	);
```

coderkevin replied:
Maybe we should change the prefix to `wc_admin` instead? That describes the usage and still namespaces it a bit. 3rd parties can still extend via the filter.

joshuatf replied:
@coderkevin Sounds like a good idea to me!

coderkevin replied:
In d40ed87, I changed the prefix to `wc_admin_`, but kept the user meta as `woocommerce_meta`. Now there's room for other stuff there, but anything that we use in wc_admin will be prefixed appropriately.

joshuatf reviewed (commented):
Code looks good and tests well across different browsers/tabs!  Nice work @coderkevin 

Couple comments about the `columnPrefsKey` prop and extensibility with the user meta fields.

justinshreve commented on lib/admin.php:
```diff
@@ -338,5 +338,73 @@ function woocommerce_embed_page_header() {
+		'user',
+		'woocommerce_meta',
+		array(
+			'get_callback'    => 'wc_admin_get_user_data_values',
```
These array items should be aligned per the coding standards. https://make.wordpress.org/core/handbook/best-practices/coding-standards/php/#indentation

coderkevin replied:
Fixed in 8ba7438

justinshreve commented on lib/admin.php:
```diff
@@ -338,5 +338,73 @@ function woocommerce_embed_page_header() {
 add_action( 'in_admin_header', 'woocommerce_embed_page_header' );
+
+/**
+ * Registers WooCommerce specific user data to the Wordpress user API.
```
[capital_P_dangit](https://codex.wordpress.org/Function_Reference/capital_P_dangit)() 😄

coderkevin replied:
Haha, I'm guilty of copypasta here, honest!

coderkevin replied:
Fixed in 8ba7438

justinshreve commented on lib/admin.php:
```diff
@@ -338,5 +338,73 @@ function woocommerce_embed_page_header() {
+}
+ add_action( 'rest_api_init', 'wc_admin_register_user_data' );
+
+ /**
```
There are some alignment issues with the docblocks for the next three functions. Looks like extra spaces.

If you run `composer install` and then `./vendor/bin/phpcs lib/admin.php` there will be some warnings about the spacing:

`Found precision alignment of 1 spaces.`

coderkevin replied:
Fixed in 8ba7438

justinshreve commented on lib/admin.php:
```diff
@@ -338,5 +338,73 @@ function woocommerce_embed_page_header() {
+	$fields  = wc_admin_get_user_data_fields();
+	$updates = array();
+	foreach ( $values as $field => $value ) {
+		if ( in_array( $field, $fields ) ) {
```
`403 | WARNING | [ ] Not using strict comparison for in_array; supply true for third argument.`

coderkevin replied:
Fixed in 8ba7438

emdashcodes reviewed (commented):
A few comments about formatting in the PHP files.

coderkevin committed:
Fix formatting/comments

These are a couple of small formatting and comment fixes.

coderkevin commented on lib/admin.php:
```diff
@@ -338,5 +338,73 @@ function woocommerce_embed_page_header() {
+		'revenue_report_columns',
+	);
+
+	return apply_filters( 'wc_admin_get_user_data_fields', $user_data_fields );
```
I can't take credit for this, it's based off [Justin's original PR.](https://github.com/woocommerce/wc-admin/pull/255/files#diff-1469878d77268810c7d104694590899dR280)

joshuatf replied:
Nice work @justinshreve !

coderkevin committed:
Replace woocommerce_meta_ prefix with wc_admin_

This changes the prefix for the meta fields for the `woocommerce_meta`
user meta array to be `wc_admin_` which better hints at the purpose of
these settings.

coderkevin commented:
This PR is updated, rebased, and ready for review once more.

joshuatf reviewed (approved):
LGTM! 👍
